creusot-contracts/src/ghost/atomic_invariant.rs

```diff
@@ -0,0 +1,90 @@
+//! Define atomic invariants.
+//!
+//! Atomic invariants are used to specify concurrent operations.
+
+use crate::{ghost::FnGhost, prelude::*};
+
+/// Creusot wrapper around [`std::sync::atomic::AtomicI32`]
+pub struct AtomicI32(::std::sync::atomic::AtomicI32);
+
+/// Does a single atomic operation, in ghost
+#[opaque]
+pub struct Committer {}
+
+/// Ownership for a [`AtomicI32`].
+#[opaque]
+pub struct AtomicI32Own {}
+
+impl AtomicI32Own {
+    /// Which atomic does this holds the ownership for
+    #[logic(opaque)]
+    pub fn id(self) -> AtomicI32 {
+        dead
+    }
+
+    /// Value currently contained in the atomic
+    #[logic(opaque)]
+    pub fn value(self) -> i32 {
+        dead
+    }
+}
+
+impl Committer {
+    /// Status of the committer
+    #[logic(opaque)]
+    pub fn shot(self) -> bool {
+        dead
+    }
+
+    /// Identity of the committer
+    ///
+    /// This is used so that we can only use the committer with the right [`AtomicOwn`].
+    #[logic(opaque)]
+    pub fn id(self) -> AtomicI32 {
+        dead
+    }
+
+    /// Value that will be written to the [`AtomicOwn`]
+    #[logic(opaque)]
+    pub fn final_value(self) -> i32 {
```

Comment:
In the case of atomic updates, this should rather be a relation between the old and the new value.